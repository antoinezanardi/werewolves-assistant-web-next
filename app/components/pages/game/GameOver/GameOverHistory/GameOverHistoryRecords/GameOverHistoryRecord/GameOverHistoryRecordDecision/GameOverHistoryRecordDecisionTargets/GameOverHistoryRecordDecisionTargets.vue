<template>
  <div
    id="game-over-history-record-decision-targets"
    class="flex flex-col gap-4 items-center justify-center"
  >
    <h4
      v-if="targetRoleNameLabel"
      id="game-over-history-record-target-role-name"
      class="text-center"
    >
      {{ targetRoleNameLabel }}
    </h4>

    <ul class="mx-auto overflow-x-auto scrollbar-thin text-center w-full whitespace-nowrap">
      <GameOverHistoryRecordDecisionTarget
        v-for="gameHistoryRecordPlayTarget in targets"
        :key="gameHistoryRecordPlayTarget.player._id"
        class="game-over-history-record-target-player inline-block"
        :target="gameHistoryRecordPlayTarget"
      />
    </ul>
  </div>
</template>

<script setup lang="ts">
import type { GameOverHistoryRecordDecisionTargetsProps } from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordDecision/GameOverHistoryRecordDecisionTargets/game-over-history-record-decision-targets.types";
import GameOverHistoryRecordDecisionTarget from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordDecision/GameOverHistoryRecordDecisionTargets/GameOverHistoryRecordDecisionTarget/GameOverHistoryRecordDecisionTarget.vue";
import type { GameHistoryRecordPlayTarget } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-target/game-history-record-play-target.class";
import { useRoleName } from "~/composables/api/role/useRoleName";

const props = defineProps<GameOverHistoryRecordDecisionTargetsProps>();

const { getDefiniteRoleNameLabel } = useRoleName();

const targets = computed<GameHistoryRecordPlayTarget[]>(() => props.gameHistoryRecord.play.targets ?? []);

const isTargetAlone = computed<boolean>(() => targets.value.length === 1);

const targetRoleNameLabel = computed<string>(() => {
  if (!isTargetAlone.value) {
    return "";
  }
  const target = targets.value[0].player;

  return getDefiniteRoleNameLabel(target.role.current, 1);
});
</script>