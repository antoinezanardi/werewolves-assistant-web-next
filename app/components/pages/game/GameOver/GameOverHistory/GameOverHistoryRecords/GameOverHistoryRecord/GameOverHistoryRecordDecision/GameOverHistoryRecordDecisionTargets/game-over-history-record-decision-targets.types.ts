import type { GameHistoryRecord } from "~/composables/api/game/types/game-history-record/game-history-record.class";

type GameOverHistoryRecordDecisionTargetsProps = {
  gameHistoryRecord: GameHistoryRecord;
};

export type { GameOverHistoryRecordDecisionTargetsProps };