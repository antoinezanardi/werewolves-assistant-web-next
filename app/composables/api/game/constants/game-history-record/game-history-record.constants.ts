const GAME_HISTORY_RECORD_PLAY_VOTING_RESULTS = [
  "sheriff-election",
  "tie",
  "death",
  "inconsequential",
  "skipped",
] as const;

const GAME_HISTORY_RECORD_PLAYER_ATTRIBUTE_ALTERATION_STATUSES = [
  "attached",
  "detached",
  "activated",
] as const;

export {
  GAME_HISTORY_RECORD_PLAY_VOTING_RESULTS,
  GAME_HISTORY_RECORD_PLAYER_ATTRIBUTE_ALTERATION_STATUSES,
};