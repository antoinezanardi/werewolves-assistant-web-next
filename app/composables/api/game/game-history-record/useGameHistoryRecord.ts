import type { GameHistoryRecord } from "~/composables/api/game/types/game-history-record/game-history-record.class";

type UseGameHistoryRecord = {
  didSourceSkipped: (gameHistoryRecord: GameHistoryRecord) => boolean;
};

function useGameHistoryRecord(): UseGameHistoryRecord {
  function didSourceSkipped(gameHistoryRecord: GameHistoryRecord): boolean {
    const {
      type,
      votes,
      targets,
      didJudgeRequestAnotherVote,
      chosenCard,
    } = gameHistoryRecord.play;

    return type === "vote" && !votes ||
      type === "target" && !targets ||
      type === "request-another-vote" && didJudgeRequestAnotherVote === undefined ||
      type === "choose-card" && !chosenCard;
  }
  return { didSourceSkipped };
}

export { useGameHistoryRecord };