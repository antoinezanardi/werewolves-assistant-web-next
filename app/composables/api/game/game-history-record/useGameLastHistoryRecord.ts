import type { MaybeRef, Ref } from "vue";
import type { GameHistoryRecordPlayerAttributeAlterationStatus } from "~/composables/api/game/types/game-history-record/game-history-record-player-attribute-alteration/game-history-record-player-attribute-alteration.types";
import type { GameHistoryRecord } from "~/composables/api/game/types/game-history-record/game-history-record.class";
import type { Game } from "~/composables/api/game/types/game.class";
import type { GameSource } from "~/composables/api/game/types/game.types";
import type { PlayerAttributeName } from "~/composables/api/game/types/players/player-attribute/player-attribute.types";
import type { Player } from "~/composables/api/game/types/players/player.class";

type UseGameLastHistoryRecord = {
  lastTargetedPlayers: Ref<Player[]>;
  doesHavePlayerAttributeAlteration: (attributeName: PlayerAttributeName, source: GameSource, status: GameHistoryRecordPlayerAttributeAlterationStatus) => boolean;
};

function useGameLastHistoryRecord(game: MaybeRef<Game>): UseGameLastHistoryRecord {
  const lastGameHistoryRecord = computed<GameHistoryRecord | null>(() => (isRef(game) ? game.value.lastGameHistoryRecord : game.lastGameHistoryRecord));

  const lastTargetedPlayers = computed<Player[]>(() => {
    if (!lastGameHistoryRecord.value?.play.targets) {
      return [];
    }
    return lastGameHistoryRecord.value.play.targets.map(target => target.player);
  });

  function doesHavePlayerAttributeAlteration(attributeName: PlayerAttributeName, source: GameSource, status: GameHistoryRecordPlayerAttributeAlterationStatus): boolean {
    if (!lastGameHistoryRecord.value?.playerAttributeAlterations) {
      return false;
    }
    return lastGameHistoryRecord.value.playerAttributeAlterations.some(playerAttributeAlteration => playerAttributeAlteration.name === attributeName &&
      playerAttributeAlteration.source === source &&
      playerAttributeAlteration.status === status);
  }
  return {
    lastTargetedPlayers,
    doesHavePlayerAttributeAlteration,
  };
}

export { useGameLastHistoryRecord };