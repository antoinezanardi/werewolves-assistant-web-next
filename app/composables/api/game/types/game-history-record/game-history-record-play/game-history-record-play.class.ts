import { Expose, plainToInstance, Type } from "class-transformer";

import { GameAdditionalCard } from "~/composables/api/game/types/game-additional-card/game-additional-card.class";
import { GameHistoryRecordPlaySource } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-source/game-history-record-play-source.class";
import { GameHistoryRecordPlayTarget } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-target/game-history-record-play-target.class";
import { GameHistoryRecordPlayVote } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-vote/game-history-record-play-vote.class";
import { GameHistoryRecordPlayVoting } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-voting/game-history-record-play-voting.class";
import type { GamePlayAction, GamePlayCause, GamePlayType } from "~/composables/api/game/types/game-play/game-play.types";
import type { RoleSide } from "~/composables/api/role/types/role.types";
import { DEFAULT_PLAIN_TO_INSTANCE_OPTIONS } from "~/utils/constants/class-transformer.constants";

class GameHistoryRecordPlay {
  @Expose()
  public type: GamePlayType;

  @Expose()
  public action: GamePlayAction;

  @Expose()
  @Type(() => GameHistoryRecordPlaySource)
  public source: GameHistoryRecordPlaySource;

  @Expose()
  public cause?: GamePlayCause;

  @Expose()
  @Type(() => GameHistoryRecordPlayTarget)
  public targets?: GameHistoryRecordPlayTarget[];

  @Expose()
  @Type(() => GameHistoryRecordPlayVote)
  public votes?: GameHistoryRecordPlayVote[];

  @Expose()
  @Type(() => GameHistoryRecordPlayVoting)
  public voting?: GameHistoryRecordPlayVoting;

  @Expose()
  public didJudgeRequestAnotherVote?: boolean;

  @Expose()
  @Type(() => GameAdditionalCard)
  public chosenCard?: GameAdditionalCard;

  @Expose()
  public chosenSide?: RoleSide;

  public static create(gameHistoryRecordPlay: GameHistoryRecordPlay): GameHistoryRecordPlay {
    return plainToInstance(GameHistoryRecordPlay, gameHistoryRecordPlay, DEFAULT_PLAIN_TO_INSTANCE_OPTIONS);
  }
}

export { GameHistoryRecordPlay };