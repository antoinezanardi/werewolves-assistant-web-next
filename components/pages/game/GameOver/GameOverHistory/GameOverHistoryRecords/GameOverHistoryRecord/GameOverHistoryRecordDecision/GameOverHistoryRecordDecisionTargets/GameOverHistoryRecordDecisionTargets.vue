<template>
  <div
    id="game-over-history-record-decision-targets"
    class="flex gap-4 items-center justify-center"
  >
    <GameOverHistoryRecordDecisionTarget
      v-for="gameHistoryRecordPlayTarget in gameHistoryRecord.play.targets"
      :key="gameHistoryRecordPlayTarget.player._id"
      class="game-over-history-record-target-player w-1/4"
      :target="gameHistoryRecordPlayTarget"
    />
  </div>
</template>

<script setup lang="ts">
import type { GameOverHistoryRecordDecisionTargetsProps } from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordDecision/GameOverHistoryRecordDecisionTargets/game-over-history-record-decision-targets.types";
import GameOverHistoryRecordDecisionTarget from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordDecision/GameOverHistoryRecordDecisionTargets/GameOverHistoryRecordDecisionTarget/GameOverHistoryRecordDecisionTarget.vue";
defineProps<GameOverHistoryRecordDecisionTargetsProps>();
</script>