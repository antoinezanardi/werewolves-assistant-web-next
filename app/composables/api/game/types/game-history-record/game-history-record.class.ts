import { Expose, plainToInstance, Type } from "class-transformer";

import { GameHistoryRecordPlay } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play.class";
import { GameHistoryRecordPlayerAttributeAlteration } from "~/composables/api/game/types/game-history-record/game-history-record-player-attribute-alteration/game-history-record-player-attribute-alteration.class";
import { GamePhase } from "~/composables/api/game/types/game-phase/game-phase.class";
import { Player } from "~/composables/api/game/types/players/player.class";
import { DEFAULT_PLAIN_TO_INSTANCE_OPTIONS } from "~/utils/constants/class-transformer.constants";

class GameHistoryRecord {
  @Expose()
  public _id: string;

  @Expose()
  public gameId: string;

  @Expose()
  public turn: number;

  @Type(() => GamePhase)
  @Expose()
  public phase: GamePhase;

  @Expose()
  public tick: number;

  @Expose()
  @Type(() => GameHistoryRecordPlay)
  public play: GameHistoryRecordPlay;

  @Expose()
  @Type(() => Player)
  public revealedPlayers?: Player[];

  @Expose()
  @Type(() => Player)
  public deadPlayers?: Player[];

  @Expose()
  @Type(() => GameHistoryRecordPlayerAttributeAlteration)
  public playerAttributeAlterations?: GameHistoryRecordPlayerAttributeAlteration[];

  @Expose()
  public createdAt: Date;

  public static create(gameHistoryRecord: GameHistoryRecord): GameHistoryRecord {
    return plainToInstance(GameHistoryRecord, gameHistoryRecord, DEFAULT_PLAIN_TO_INSTANCE_OPTIONS);
  }
}

export { GameHistoryRecord };