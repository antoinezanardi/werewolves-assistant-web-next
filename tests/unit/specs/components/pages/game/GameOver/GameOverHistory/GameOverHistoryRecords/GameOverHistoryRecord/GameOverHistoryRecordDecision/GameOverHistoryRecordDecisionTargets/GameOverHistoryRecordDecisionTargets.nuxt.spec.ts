import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";

import { createFakeGameHistoryRecordPlayTarget } from "@tests/unit/utils/factories/composables/api/game/game-history-record/game-history-record-play/game-history-record-play-target/game-history-record-play-target.factory";
import { createFakeGameHistoryRecordPlay } from "@tests/unit/utils/factories/composables/api/game/game-history-record/game-history-record-play/game-history-record-play.factory";
import { createFakeGameHistoryRecord } from "@tests/unit/utils/factories/composables/api/game/game-history-record/game-history-record.factory";
import { createFakeActorAlivePlayer } from "@tests/unit/utils/factories/composables/api/game/player/player-with-role.factory";
import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type { GameOverHistoryRecordDecisionTargetsProps } from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordDecision/GameOverHistoryRecordDecisionTargets/game-over-history-record-decision-targets.types";
import type GameOverHistoryRecordDecisionTarget from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordDecision/GameOverHistoryRecordDecisionTargets/GameOverHistoryRecordDecisionTarget/GameOverHistoryRecordDecisionTarget.vue";
import GameOverHistoryRecordDecisionTargets from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordDecision/GameOverHistoryRecordDecisionTargets/GameOverHistoryRecordDecisionTargets.vue";
import type { GameHistoryRecordPlayTarget } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-target/game-history-record-play-target.class";

describe("Game Over History Record Decision Targets Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameOverHistoryRecordDecisionTargets>>;
  const defaultPlayers = [
    createFakeActorAlivePlayer({ name: "Antoine" }),
    createFakeActorAlivePlayer({ name: "Olivia" }),
    createFakeActorAlivePlayer({ name: "Thomas" }),
    createFakeActorAlivePlayer({ name: "Doudou" }),
  ];
  const defaultTargets = [
    createFakeGameHistoryRecordPlayTarget({ player: defaultPlayers[0] }),
    createFakeGameHistoryRecordPlayTarget({ player: defaultPlayers[1] }),
    createFakeGameHistoryRecordPlayTarget({ player: defaultPlayers[2] }),
    createFakeGameHistoryRecordPlayTarget({ player: defaultPlayers[3] }),
  ];
  const defaultGameHistoryRecord = createFakeGameHistoryRecord({
    play: createFakeGameHistoryRecordPlay({
      type: "target",
      action: "eat",
      targets: defaultTargets,
    }),
  });
  const defaultProps: GameOverHistoryRecordDecisionTargetsProps = { gameHistoryRecord: defaultGameHistoryRecord };

  async function mountGameOverHistoryRecordDecisionTargetsComponent(options: ComponentMountingOptions<typeof GameOverHistoryRecordDecisionTargets> = {}):
  Promise<ReturnType<typeof mount<typeof GameOverHistoryRecordDecisionTargets>>> {
    return mountSuspendedComponent(GameOverHistoryRecordDecisionTargets, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameOverHistoryRecordDecisionTargetsComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Target Role Title", () => {
    it("should render the target role title when there only one target.", async() => {
      wrapper = await mountGameOverHistoryRecordDecisionTargetsComponent({
        props: {
          gameHistoryRecord: createFakeGameHistoryRecord({
            play: createFakeGameHistoryRecordPlay({
              type: "target",
              action: "eat",
              targets: [
                createFakeGameHistoryRecordPlayTarget({
                  player: createFakeActorAlivePlayer(),
                }),
              ],
            }),
          }),
        },
      });
      const targetRoleTitle = wrapper.find<HTMLHeadingElement>("#game-over-history-record-target-role-name");

      expect(targetRoleTitle.text()).toBe("shared.role.definiteName.actor, 1");
    });

    it("should not render the target role title when there are multiple targets.", () => {
      const targetRoleTitle = wrapper.find<HTMLHeadingElement>("#game-over-history-record-target-role-name");

      expect(targetRoleTitle.exists()).toBeFalsy();
    });
  });

  describe("Targets", () => {
    it("should render the targets when rendered.", () => {
      const targets = wrapper.findAllComponents<typeof GameOverHistoryRecordDecisionTarget>(".game-over-history-record-target-player");

      expect(targets).toHaveLength(4);
      expect(targets[0].props("target")).toStrictEqual<GameHistoryRecordPlayTarget>(defaultTargets[0]);
      expect(targets[1].props("target")).toStrictEqual<GameHistoryRecordPlayTarget>(defaultTargets[1]);
      expect(targets[2].props("target")).toStrictEqual<GameHistoryRecordPlayTarget>(defaultTargets[2]);
      expect(targets[3].props("target")).toStrictEqual<GameHistoryRecordPlayTarget>(defaultTargets[3]);
    });

    it("should not render the targets when there are no targets.", async() => {
      const gameHistoryRecord = createFakeGameHistoryRecord({
        play: createFakeGameHistoryRecordPlay({
          type: "target",
          action: "eat",
        }),
      });
      wrapper = await mountGameOverHistoryRecordDecisionTargetsComponent({ props: { gameHistoryRecord } });
      const targets = wrapper.findAllComponents<typeof GameOverHistoryRecordDecisionTarget>(".game-over-history-record-target-player");

      expect(targets).toHaveLength(0);
    });
  });
});