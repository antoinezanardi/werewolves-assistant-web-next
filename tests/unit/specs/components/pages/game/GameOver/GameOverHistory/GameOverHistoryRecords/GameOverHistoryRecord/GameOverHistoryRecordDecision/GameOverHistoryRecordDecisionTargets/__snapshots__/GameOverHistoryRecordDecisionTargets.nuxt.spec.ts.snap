// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Over History Record Decision Targets Component > should match snapshot when rendered. 1`] = `
"<div id="game-over-history-record-decision-targets" class="flex flex-col gap-4 items-center justify-center">
  <!--v-if-->
  <ul class="mx-auto overflow-x-auto scrollbar-thin text-center w-full whitespace-nowrap">
    <game-over-history-record-decision-target-stub target="[object Object]" class="game-over-history-record-target-player inline-block"></game-over-history-record-decision-target-stub>
    <game-over-history-record-decision-target-stub target="[object Object]" class="game-over-history-record-target-player inline-block"></game-over-history-record-decision-target-stub>
    <game-over-history-record-decision-target-stub target="[object Object]" class="game-over-history-record-target-player inline-block"></game-over-history-record-decision-target-stub>
    <game-over-history-record-decision-target-stub target="[object Object]" class="game-over-history-record-target-player inline-block"></game-over-history-record-decision-target-stub>
  </ul>
</div>"
`;
